### PR TITLE
Feature/kup

### PR DESCRIPTION
**Pull request by dkcumming**
Change `kmir` `nix` system so that it is `kup` compatible

### CODE REVIEW

**Comment by dkcumming** [comment]
@yanliu18 Nix able to build `kmir` just need to figure out how to provide the definitions if it were build with `kup`, as `$(poetry run kbuild which [llvm | haskell])` will not work

File: kmir/src/kmir/__main__.py [hunk 1]
```diff
@@ -83,6 +95,7 @@ def create_argument_parser() -> ArgumentParser:
     )
     parse_subparser.add_argument(
         '--definition-dir',
+        default=None,
```
Comment (dkcumming) [inline kmir/src/kmir/__main__.py:98]: @yanliu18 `--definition` could be omitted after this change, but it won't work unless it's a nix build. I will get it to work locally in the `feature/default-definition-paths` branch.

File: kmir/src/kmir/kmir.py [hunk 2]
```diff
@@ -25,15 +26,38 @@ class KMIR:
...
 
-        haskell_dir = Path(haskell_dir)
+        if haskell_dir is None:
+            env_haskell_dir = os.getenv('KMIR_HASKELL_DIR')
+            if env_haskell_dir:
+                haskell_dir = Path(env_haskell_dir)
+            else:
+                raise RuntimeError(
+                    'Cannot find KMIR HASKELL definition, please specify --definition-dir, or KMIR_HASKELL_DIR'
+                )
```
Comment (goodlyrottenapple) [inline kmir/src/kmir/kmir.py:58]: should this be an error given the haskell definition is unused atm (as far as i understand)?

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+              mkdir -p $out/bin/
+            '';
+          };
+        })
+        # Now mir-semantics is built, wrap kmir with LLVM and HASKELL defs
+        (final: prev: {
+          kmir = prev.kmir.overrideAttrs (oldAttrs: {
+            buildInputs = oldAttrs.buildInputs or [] ++ [ prev.python310 prev.makeWrapper ];
+            postInstall = oldAttrs.postInstall or "" + ''
+              wrapProgram $out/bin/kmir --set KMIR_LLVM_DIR ${(toString prev.mir-semantics) + "/lib/llvm"} --set KMIR_HASKELL_DIR ${(toString prev.mir-semantics) + "/lib/haskell"} --prefix PATH : ${prev.lib.makeBinPath [ k-framework.packages.${prev.system}.k ]}
```
Comment (goodlyrottenapple) [inline flake.nix:85]:
```suggestion
              wrapProgram $out/bin/kmir --set KMIR_LLVM_DIR "$out/lib/llvm" --set KMIR_HASKELL_DIR "$out/lib/haskell" --prefix PATH : ${prev.lib.makeBinPath [ k-framework.packages.${prev.system}.k ]}
```

bad things usually happen if a derivation tries to refer to itself via anything other than the`$out` variable.

File: kmir/src/kmir/kmir.py (hunk 2 above)
Comment (dkcumming) [reply kmir/src/kmir/kmir.py:58]: That's a good point, for now I will just print instead of raising an error

**Review by goodlyrottenapple — COMMENTED** [review]
LGTM on the nix side of things

File: flake.nix (hunk 3 above)
Comment (dkcumming) [reply flake.nix:85]: I couldn't get this to work previously, and I tried this change just now, it doesn't route correctly. I think it `$out` is the python kmir dir, and that is not where the defs are:
```
$ kmir parse src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir 
Traceback (most recent call last):
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/bin/..kmir-wrapped-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/__main__.py", line 22, in main
    execute(**vars(args))
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/__main__.py", line 39, in exec_parse
    kmir = KMIR(definition_dir, None)
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/kmir.py", line 45, in __init__
    check_dir_path(llvm_dir)
  File "/nix/store/b0197fwlh6szlhv1qsh7n99baz8lvvgc-python3.10-pyk-0.1.401/lib/python3.10/site-packages/pyk/utils.py", line 420, in check_dir_path
    raise ValueError(f'Directory does not exist: {path}')
ValueError: Directory does not exist: /nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/llvm
```

Comment (dkcumming) [reply flake.nix:85]: Unless I am mistaken, we want this result:
`export KMIR_LLVM_DIR='/nix/store/3rbbjf78r362vam1gykb3nw37h754nz0-mir-semantics-a386aad115d93c92d2b1418c33086833ef17a098/lib/llvm'`
Not this result:
`export KMIR_LLVM_DIR='/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/llvm'`

Is there someway I can achieve that without referencing `prev.mir-semantics`? I couldn't get it to work.

**Review by goodlyrottenapple — APPROVED** [review]

**Pull request merged**